pkg/scheduler/queue/user_queues.go

```diff
@@ -28,298 +27,350 @@ type querier struct {
 	disconnectedAt time.Time
 }
 
-// This struct holds user queues for pending requests. It also keeps track of connected queriers,
-// and mapping between users and queriers.
-type queues struct {
-	userQueues map[string]*userQueue
-
-	// List of all users with queues, used for iteration when searching for next queue to handle.
-	// Users removed from the middle are replaced with "". To avoid skipping users during iteration, we only shrink
-	// this list when there are ""'s at the end of it.
-	users []string
-
-	maxUserQueueSize int
+type tenantQuerierState struct {
+	// a tenant has many queriers
+	// a tenant has *all* queriers if:
+	//  - sharding is disabled (max-queriers-per-tenant=0)
+	//  - or if max-queriers-per-tenant >= the number of queriers
+	//
+	// Tenant -> Queriers is the core relationship randomized from the shuffle shard seed.
+	// The shuffle shard seed is itself consistently hashed from the tenant ID.
+	// However, the most common operation is the querier asking for its next request,
+	// which requires a relatively efficient lookup or check of Querier -> Tenant.
+	//
+	// Reshuffling is done when:
+	//  - a querier connection is added or removed
+	//  - it is detected during request enqueueing that a tenant's queriers
+	//    were calculated from an outdated max-queriers-per-tenant value
+
+	queriersByID map[string]*querierConn
```

Comment:
Would be good to clarify which ID we're referring to as the key here.

We have a couple of different IDs stored as strings throughout this file - perhaps adding some specific type definitions would help clarify which one is what and also catch places where we use the wrong kind of ID?

Reply:
it's a bit verbose because we have to alias search and sort implementations for the slice of querierIDs when we make that a type alias, but I have a good solution - will hold it until your renaming PR gets in because I have to change a bunch of function signatures

Reply:
Makes sense, thanks.